kvserver: add Raft transport queue byte size metric

pav-kv opened the pull request:
The new metric measures the total size in bytes of the outgoing Raft messages
queue in `RaftTransport`. In the future it could be used for rate-limiting.

Experiment running a local cluster with added 10ms delay on queue receiving end:
<img width="996" alt="Screenshot 2022-08-30 at 14 12 55" src="https://user-images.githubusercontent.com/3757441/187446097-9c8bb933-7918-4922-a4f9-55fb9cd74e6d.png">

Fixes #86371

Release justification: helps investigating customer issues
Release note: None

pav-kv committed:
kvserver: add Raft transport queue byte size metric

The new metric measures the total size in bytes of the outgoing Raft messages
queue in RaftTransport. In the future it could be used for rate-limiting.

Release justification: helps investigating customer issues
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86942)
<!-- Reviewable:end -->

pavelkalinnikov commented on pkg/kv/kvserver/raft_transport.go:
```diff
@@ -534,19 +548,20 @@ func (t *RaftTransport) startProcessNewQueue(
-			case <-ch:
+			case req := <-q.reqs:
+				atomic.AddInt64(&q.bytes, -int64(req.Size()))
 				t.metrics.MessagesDropped.Inc(1)
```
Should we `releaseRaftMessageRequest(req)` here too?

tbg reviewed (commented):
Just a quick glance in with some comments, but I think this looks good and it won't take long to get this merged last week.
Re: my suggested changes, I *think* they wouldn't prevent us from backporting to 22.2, but keep in mind that we potentially do want to do that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-:-NAPmE7p-xlhtAq4V-7H:bc7hsu2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov and @tbg)

___
*[`pkg/kv/kvserver/raft_transport.go` line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPjCeyBBhbdWb-Flbf:-NAPjCeyBBhbdWb-Flbg:bnw9os6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L167)):*
> ```Go
> type raftMessageQueue struct {
> 	reqs  chan *kvserverpb.RaftMessageRequest
> 	bytes int64 // the number of bytes in flight
> ```

Let's call this `raftSendQueue` so that it's clear which way the messages are queued. (We already call the converse `raftReceiveQueue`).

Since the `bytes` field has special access rules (it's an atomic) what I would do is this:

```go
type raftSendQueue struct {
  _bytes int64 // atomic; use `addBytes()` to access
}

func (q *raftSendQueue) addBytes(n int64) {
  return atomic.AddInt64(&q._bytes, n)
}
```
___
*[`pkg/kv/kvserver/raft_transport.go` line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPl-IOBrv521PB3kir:-NAPl-IOBrv521PB3kis:b-vpriai) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L439)):*
> ```Go
> 			for budget > 0 {
> 				select {
> 				case req = <-q.reqs:
> ```

I'm noticing that we pull off `q.reqs` in a few places, but none of them have more than one channel they listen on. So we could unify:

```go
for budget > 0 {
  req, ok := q.tryDequeue()
  if !ok {
    break
  }
}
```

and similar in other places. `tryDequeue` could then manage `q.bytes` and we'd have a cleaner contract for how to consume from a queue.

We can similarly have a `tryEnqueue`, I think this makes sense even though it'd only be used in one place. Whatever we can comfortably encapsulate in `raftSendQueue`, I think we should.

Btw, even if we need to have a truly blocking dequeue or enqueue, we should still be OK because we can have a `dequeue` flavor that takes an additional channel (or two, or three, etc), so I don't think of this leading us down a potential cul de sac.
___
*[`pkg/kv/kvserver/raft_transport.go` line 519 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPkqrB8E7OWX4S4uTK:-NAPkqrB8E7OWX4S4uTL:b-4457h2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L519)):*
> ```Go
> 	select {
> 	case q.reqs <- req:
> 		atomic.AddInt64(&q.bytes, int64(req.Size()))
> ```

`q.addBytes(int64(req.Size())`
___
*[`pkg/kv/kvserver/raft_transport.go` line 525 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPkmh54xpC_0z-IgtJ:-NAPkmh54xpC_0z-IgtK:b-swt8ky) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L525)):*
> ```Go
> 			log.Warningf(t.AnnotateCtx(context.Background()), "raft send queue to n%d is full", toNodeID)
> 		}
> 		releaseRaftMessageRequest(req) // TODO(pavelkalinnikov): dropped++
> ```

Reminder to address this TODO.
___
*[`pkg/kv/kvserver/raft_transport.go` line 553 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPQhhnFCnyrwYdpSM3-r1-553:-NAPkQiR8z5CulSfJgYk:bash55k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L553)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Should we `releaseRaftMessageRequest(req)` here too?
</blockquote></details>

Let's do it for completeness, even though it shouldn't matter.
___
*[`pkg/kv/kvserver/raft_transport_metrics.go` line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPlxK_1dJfCmpUPu0b:-NAPlxK_1dJfCmpUPu0c:b-4v10r4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport_metrics.go#L47)):*
> ```Go
> 
> The queue is composed of multiple bounded channels associated with different
> peers. The overall size of tens of MiB could indicate issues streaming messages
> ```

When we replicate AddSST, we have individual entries that are > 16mb, so this heuristic wouldn't hold up so well. Maybe better here to say that either a high queue size or queue length in the thousands is a possible reason for concern, especially if it's a significant difference to baseline.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov and @tbg)

___
*[`pkg/kv/kvserver/raft_transport.go` line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPjCeyBBhbdWb-Flbf:-NAPo2fNE-sk36E7KV1s:bdfox1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L167)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Let's call this `raftSendQueue` so that it's clear which way the messages are queued. (We already call the converse `raftReceiveQueue`).

Since the `bytes` field has special access rules (it's an atomic) what I would do is this:

```go
type raftSendQueue struct {
  _bytes int64 // atomic; use `addBytes()` to access
}

func (q *raftSendQueue) addBytes(n int64) {
  return atomic.AddInt64(&q._bytes, n)
}
```
</blockquote></details>

Btw, you just helped my to find a bug :) I forgot to `Load` on the reading end.

Replaced the type of this field by an ad-hoc atomic int64 wrapper which doesn't automatically convert to int64, so it must be safe enough. Also, found it not necessary to add underscore to the field name, as it looks non-typical for Go (are you referring to some convention here?).

Do you know when we are going to switch to Go 1.19? If it's not too soon, we could turn this ad-hoc type into a proper one, and convert the whole package/repo to use it. Potentially it would reveal similar bugs to mine, like forgetting that the variable should be accessed using atomics. I see a few [places](https://github.com/cockroachdb/cockroach/search?q=atomic.Int64) across the repo where apparently the variable name is used as a reminder :)
___
*[`pkg/kv/kvserver/raft_transport.go` line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPl-IOBrv521PB3kir:-NAPqaz90YLjaT8ZkkDR:bnz39rq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L439)):*
I agree that some encapsulation would be nice here, but:

> none of them have more than one channel they listen on

There is one multi-channel case a few lines above, with 4 channels. I am reluctant to adding a method with 4 channels here. If this method receives something through one of the 3 other channels, it has to do something about it, so we would need to essentially copy most of the body of this `select` there, or find a way to return a struct that represents one of the 4 outcomes.

I could still add the non-blocking methods, and have one special place where the abstraction leaks. OTOH I would prefer consistency here.

One more minor argument against "try" methods is that `req.Size()` would need to be computed twice, which seems unnecessary.

Zooming out a bit, the whole sending part of `processQueue` (which needs to be split in 2 really) could be thought of `raftSendQueue` method, so it's okay that it knows how to use its fields.
___
*[`pkg/kv/kvserver/raft_transport.go` line 519 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPkqrB8E7OWX4S4uTK:-NAQ1f_cBQNUc65mnG-T:bbm6cv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L519)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

`q.addBytes(int64(req.Size())`
</blockquote></details>

replaced by q.bytes.add()
___
*[`pkg/kv/kvserver/raft_transport.go` line 525 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPkmh54xpC_0z-IgtJ:-NAQ1pgQ9XUb62uD7tfY:bm9hyu6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L525)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Reminder to address this TODO.
</blockquote></details>

I was going to do it in a separate PR. Do you think it's ok to pack in this one?
___
*[`pkg/kv/kvserver/raft_transport.go` line 553 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPQhhnFCnyrwYdpSM3-r1-553:-NAQ1ymwAJcLutyxhccY:bggvzrv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L553)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Let's do it for completeness, even though it shouldn't matter.
</blockquote></details>

Same question: is it ok to put in the same PR?


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> Reviewed 2 of 4 files at r1, 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov and @tbg)

___
*[`pkg/kv/kvserver/raft_transport.go` line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPjCeyBBhbdWb-Flbf:-NAd9Yy7-S2CfKqwDD3b:bhagpf7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L167)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Btw, you just helped my to find a bug :) I forgot to `Load` on the reading end.

Replaced the type of this field by an ad-hoc atomic int64 wrapper which doesn't automatically convert to int64, so it must be safe enough. Also, found it not necessary to add underscore to the field name, as it looks non-typical for Go (are you referring to some convention here?).

Do you know when we are going to switch to Go 1.19? If it's not too soon, we could turn this ad-hoc type into a proper one, and convert the whole package/repo to use it. Potentially it would reveal similar bugs to mine, like forgetting that the variable should be accessed using atomics. I see a few [places](https://github.com/cockroachdb/cockroach/search?q=atomic.Int64) across the repo where apparently the variable name is used as a reminder :)
</blockquote></details>

We've been attempting a 1.19 bump before the 22.2 branch cut, but looks like it ran into some issues: #86049. Not sure if it'll land in time.

I think I'd prefer just using a plain ol' int64 with atomic access, since that's what we usually do, but no strong opinion either way. Looking forward to `atomic.Int64`.
___
*[`pkg/kv/kvserver/raft_transport.go` line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPl-IOBrv521PB3kir:-NAdDS_1EnWLwyNyIHHy:b-b2wz0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L439)):*
> There is one multi-channel case a few lines above, with 4 channels. I am reluctant to adding a method with 4 channels here. 

Yeah, I think it's fine to keep the channel and require readers to update `bytes`, since it's only used in a few places internally in this package/file.
___
*[`pkg/kv/kvserver/raft_transport.go` line 525 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPkmh54xpC_0z-IgtJ:-NAdE6bt0zxOrksM0m8G:b3yzl1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L525)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

I was going to do it in a separate PR. Do you think it's ok to pack in this one?
</blockquote></details>

Separate commit would be fine, but a separate PR works too.
___
*[`pkg/kv/kvserver/raft_transport.go` line 212 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAdCNyP-tqvgJcZ-e1p:-NAdCNyP-tqvgJcZ-e1q:b-twjaft) ([raw file](https://github.com/cockroachdb/cockroach/blob/4173d859e1f1d5add4397f965badee34417c7d40/pkg/kv/kvserver/raft_transport.go#L212)):*
> ```Go
> // aggQueueMetric returns a functional gauge metric which computes the sum of
> // the passed-in metric function over all sub-queues.
> func (t *RaftTransport) aggQueueMetric(metric func(*raftSendQueue) int64) func() int64 {
> ```

nit: slight preference for just adding two separate functions for this, possibly with an internal visitor helper, so that we can keep the logic internal. This doesn't generalize all that well in any case, since it's limited to sums.


<!-- Sent from Reviewable.io -->

pavelkalinnikov commented on pkg/kv/kvserver/raft_transport.go:
```diff
@@ -502,13 +517,14 @@ func (t *RaftTransport) SendAsync(
 	select {
-	case ch <- req:
+	case q.reqs <- req:
+		q.bytes.add(int64(req.Size()))
```
I think there is a subtle bug possibility here.

It looks like requests are “released” into the pool (I guess, to avoid allocations). So, after we send `req`, it can be written to concurrently, hence it's not safe/correct to call `req.Size()` like this.

I'll put the `req.Size()` computation before the send, with a comment why.

erikgrinaker replied:
Makes sense. This would also make sense if we were to start using these for memory budgeting, where we would have to allocate budget before the send, and free it after the receive.

pavelkalinnikov replied:
I actually did hit this bug while testing locally. I observed the metric going negative. Probably what happened is that sometimes the "release" overtook and emptied the request, so `Add(req.Size())` was leaky (and reported a nearly-0 instead of real size), unlike the accurate `Add(-req.Size())` on the receiving end.

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @pavelkalinnikov, and @tbg)

___
*[`pkg/kv/kvserver/raft_transport.go` line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPjCeyBBhbdWb-Flbf:-NAiX8g76sOKPhGMbtrl:b-4y3ip1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L167)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

We've been attempting a 1.19 bump before the 22.2 branch cut, but looks like it ran into some issues: #86049. Not sure if it'll land in time.

I think I'd prefer just using a plain ol' int64 with atomic access, since that's what we usually do, but no strong opinion either way. Looking forward to `atomic.Int64`.
</blockquote></details>

Reverted to plain old ints. Commented the field that it should be accessed **atomically**, and left a TODO to use the type-safe `atomic.Int64` in Go 1.19+. Hope this suffices.

I am reluctant to introduce `add/load` methods to `raftSendQueue` for the sake of accessing one field, I think it should be baked into the type of this field rather than the containing type. Hopefully the TODO won't sit there long. I'll file an issue to convert this and other places to the type-safe `atomic`s.
___
*[`pkg/kv/kvserver/raft_transport.go` line 439 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPl-IOBrv521PB3kir:-NAiQOGK5uWjT7BaWrCY:b-2mpiwm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L439)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

> There is one multi-channel case a few lines above, with 4 channels. I am reluctant to adding a method with 4 channels here. 

Yeah, I think it's fine to keep the channel and require readers to update `bytes`, since it's only used in a few places internally in this package/file.
</blockquote></details>

Added a comment in the struct saying that the metric must be updated on each send/recv, hopefully that suffices.
___
*[`pkg/kv/kvserver/raft_transport.go` line 519 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPkqrB8E7OWX4S4uTK:-NAiYHeLEy0MSmadpr0_:baghknm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L519)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

replaced by q.bytes.add()
</blockquote></details>

reverted to plain old ints again, after Erik's comment
___
*[`pkg/kv/kvserver/raft_transport.go` line 212 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAdCNyP-tqvgJcZ-e1p:-NAiWCibDuTCcVQsLiDg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4173d859e1f1d5add4397f965badee34417c7d40/pkg/kv/kvserver/raft_transport.go#L212)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: slight preference for just adding two separate functions for this, possibly with an internal visitor helper, so that we can keep the logic internal. This doesn't generalize all that well in any case, since it's limited to sums.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/raft_transport_metrics.go` line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPlxK_1dJfCmpUPu0b:-NAiTeKSEix2MP6VXFnN:bm34pc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport_metrics.go#L47)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

When we replicate AddSST, we have individual entries that are > 16mb, so this heuristic wouldn't hold up so well. Maybe better here to say that either a high queue size or queue length in the thousands is a possible reason for concern, especially if it's a significant difference to baseline.
</blockquote></details>

Reworded a bit. Looks better?


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

___
*[`pkg/kv/kvserver/raft_transport.go` line 525 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPkmh54xpC_0z-IgtJ:-NAiuCWbDX8bmnNCmAs_:b-lr92d4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L525)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Separate commit would be fine, but a separate PR works too.
</blockquote></details>

Ah, this is not needed, because there is a `defer` in the beginning of this method, which will increase the metric.


<!-- Sent from Reviewable.io -->

pav-kv committed:
kvserver: release dropped Raft messages into the pool

Fix a couple of places where the dropped Raft messages don't get returned to
the pool.

Release justification: small bug fix
Release note: None

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

___
*[`pkg/kv/kvserver/raft_transport.go` line 553 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPQhhnFCnyrwYdpSM3-r1-553:-NAivO8yDH3J_ne4nPNK:b-ejt3uu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport.go#L553)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Same question: is it ok to put in the same PR?
</blockquote></details>

Added a commit with a couple of message release fixes.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> Reviewed 1 of 4 files at r1, 3 of 3 files at r4, 1 of 1 files at r5, all commit messages.
> Dismissed @pavelkalinnikov from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/raft_transport_metrics.go` line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPlxK_1dJfCmpUPu0b:-NAnEhPX0f2Py5oqJc3R:b6uhopg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport_metrics.go#L47)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Reworded a bit. Looks better?
</blockquote></details>

Thanks! nit: trailing dot needed I think but not worth another trip through CI.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

___
*[`pkg/kv/kvserver/raft_transport_metrics.go` line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-NAPlxK_1dJfCmpUPu0b:-NAnSOiK2JPpQjgFRvWm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce4b146370be07506c4bc1cbb4a6fefcd0d36614/pkg/kv/kvserver/raft_transport_metrics.go#L47)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Thanks! nit: trailing dot needed I think but not worth another trip through CI.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 3 files at r4, 2 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86942#-:-NAneLwNCHb23tH7m-Sm:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @pavelkalinnikov)



<!-- Sent from Reviewable.io -->

pavelkalinnikov commented:
bors r=tbg,erikgrinaker

pavelkalinnikov commented:
blathers backport 22.1

tbg commented:
I think we can give up on the backport. When the queue is large and causes OOMs, this is usually captured in profiles. The main value of the metric is as a stepping stone for actual budgeting, and for experiments, which we'll run on master.
